minifi/minifi-bootstrap/src/main/java/org/apache/nifi/minifi/bootstrap/util/OSUtils.java

```diff
@@ -0,0 +1,152 @@
+     * the platform specific ProcessId for Unix like systems or Handle for Win32 Systems, a.k.a <b>pid</b>
+     * In-case it fails to determine the pid, it will return Null.
+     * Purpose for the Logger is to log any interaction for debugging.
+     */
+    public static Long getProcessId(final Process process, final Logger logger) {
+        /*
+         * NIFI-5175: NiFi built with Java 1.8 and running on Java 9.  Reflectively invoke Process.pid() on
+         * the given process instance to get the PID of this Java process.  Reflection is required in this scenario
+         * due to NiFi being compiled on Java 1.8, which does not have the Process API improvements available in
+         * Java 9.
+         *
+         * Otherwise, if NiFi is running on Java 1.8, attempt to get PID using capabilities available on Java 1.8.
+         *
+         * TODO: When minimum Java version updated to Java 9+, this class should be removed with the addition
+         * of the pid method to the Process API.
+         */
+        Long pid = null;
+        try {
+            // Get Process.pid() interface method to avoid illegal reflective access
+            final Method pidMethod = Process.class.getDeclaredMethod("pid");
+            final Object pidNumber = pidMethod.invoke(process);
+            if (pidNumber instanceof Long) {
+                pid = (Long) pidNumber;
+            }
+        } catch (final NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
+            final String processClassName = process.getClass().getName();
+            if (processClassName.equals("java.lang.UNIXProcess")) {
+                pid = getUnixPid(process, logger);
+            } else if (processClassName.equals("java.lang.Win32Process")
+                    || processClassName.equals("java.lang.ProcessImpl")) {
+                pid = getWindowsProcessId(process, logger);
+            } else {
+                logger.debug("Failed to determine Process ID from [{}]: {}", processClassName, e.getMessage());
+            }
+        }
+
+        return pid;
+    }
+
+    /**
+     * Returns the major version parsed from the provided Java version string (e.g. {@code "1.8.0.231"} -> {@code 8}).
+     *
+     * @param version the Java version string
+     * @return the major version as an int
+     */
+    public static int parseJavaVersion(final String version) {
+        String majorVersion;
+        if (version.startsWith("1.")) {
+            majorVersion = version.substring(2, 3);
+        } else {
+            Pattern majorVersion9PlusPattern = Pattern.compile("(\\d+).*");
+            Matcher m = majorVersion9PlusPattern.matcher(version);
+            if (m.find()) {
+                majorVersion = m.group(1);
+            } else {
+                throw new IllegalArgumentException("Could not detect major version of " + version);
+            }
+        }
+        return Integer.parseInt(majorVersion);
+    }
```

Comment:
This method can be removed since it is not used for Process ID determination.

Reply:
I waffled on whether to make a new module, but I think you're right. I can create a "nifi-bootstrap-utils" module and just include this for now (and share it among nifi-bootstrap and minifi-bootstrap). That way the code is in one place to maintain, but doesn't add any footprint as it would if it were in nifi-utils.

Reply:
Thanks, that sounds like a good way forward.